### ISSUE TITLE
Excessive calls to OTP code_server process

### ISSUE THREAD

**Issue by slfritchie** [basho/riak_cs#669]
This is a performance bug, noticed while using Riak CS as a test subject for exploring Erlang profiling and tracing tools.  The items below are using Basho's Riak CS 1.4.1 package for OS X.  I am using a developer version of Riak, I hope that doesn't matter.

When looking at the eprof profile of Riak CS, measured across all processes for 10 seconds, while a single load generator process is running:

```
sh -c 'while true; do s3cmd la | wc -l; done'
```

... which lists 992 objects, so more than zero objects and less than a zillion.  All of the eprof functions which use 1% or more time are:

```
FUNCTION                                                                     CALLS      %    TIME  [uS / CALLS]
--------                                                                     -----    ---    ----  [----------]
[...]
error_handler:check_inheritance/2                                            97613   1.00   29201  [      0.30]
erlang:whereis/1                                                             97754   1.00   29263  [      0.30]
erlang:list_to_binary/1                                                      39690   1.06   31025  [      0.78]
string:tokens2/4                                                            240430   1.13   32828  [      0.14]
lists:reverse/2                                                              23978   1.13   32917  [      1.37]
xmerl:export_element/2                                                       97589   1.16   33909  [      0.35]
erlang:binary_to_term/1                                                        753   1.19   34531  [     45.86]
error_handler:ensure_loaded/1                                                97612   1.21   35382  [      0.36]
xmerl_lib:end_tag/1                                                         195184   1.23   35825  [      0.18]
string:tokens/2                                                              10050   1.30   37884  [      3.77]
riak_cs_utils:binary_to_hexlist/1                                           168470   1.34   39047  [      0.23]
xmerl_lib:start_tag/2                                                       195184   1.40   40785  [      0.21]
erlang:atom_to_list/1                                                       195440   1.40   40798  [      0.21]
lists:member/2                                                              278415   1.44   41890  [      0.15]
erlang:raise/3                                                               97613   1.56   45576  [      0.47]
code_server:handle_call/3                                                    97613   1.67   48790  [      0.50]
riak_cs_utils:num2hexchar/1                                                 317120   1.72   50059  [      0.16]
code_server:loop/1                                                           97613   1.95   56728  [      0.58]
erlang:get_module_info/2                                                     97653   2.06   60133  [      0.62]
unicode:characters_to_list/2                                                 59480   2.11   61452  [      1.03]
code_server:reply/2                                                          97613   2.16   62871  [      0.64]
xmerl_lib:expand_content/4                                                  248170   2.21   64462  [      0.26]
gen_fsm:loop/7                                                                 975   2.24   65157  [     66.83]
lists:merge2_2/5                                                            300176   2.25   65614  [      0.22]
erlang:error/1                                                               97613   2.32   67523  [      0.69]
xmerl:export_content/2                                                      271087   2.36   68827  [      0.25]
xmerl_lib:expand_element/4                                                  158810   2.68   78173  [      0.49]
xmerl:apply_cb/5                                                            368682   2.74   79902  [      0.22]
unicode:characters_to_binary/2                                                  65   3.69  107526  [   1654.25]
code_server:call/2                                                           97612   5.00  145653  [      1.49]
erlang:get_stacktrace/0                                                      97613   9.91  288793  [      2.96]
xmerl_lib:export_text/2                                                    1656643  13.27  386699  [      0.23]
```

9.7K calls per second to the code_server is a bit excessive.  The fact that the code server is being called at all during normal operations is something that needs be eliminated, IMHO.

When I use this `redbug` call, I see a 1 call to load `riak_cs_wm_common`, which feels wrong to me.  However, the 99 other calls are to load `xmerl_xml`.  WTF, that's horrible.

```
redbug:start("code_server:call", [{msgs,100},{max_msg_size, 9999999999999}]).
```

Call stack detail (sample size of _one_):

```
(riak-cs@127.0.0.1)13> redbug:start("code_server:call -> stack", {max_msg_size, 9999999999999}).    
ok

16:29:31 <{mochiweb_acceptor,init,3}> {code_server,call,
                                       [code_server,
                                        {ensure_loaded,xmerl_xml}]}
  "proc_lib:init_p_do_apply/3 "
  "mochiweb_http:parse_headers/5 "
  "webmachine_mochiweb:loop/2 "
  "webmachine_decision_core:handle_request/2 "
  "webmachine_decision_core:decision/1 "
  "webmachine_decision_core:resource_call/1 "
  "webmachine_resource:do/3 "
  "webmachine_resource:resource_call/3 "
  "riak_cs_wm_common:produce_body/2 "
  "riak_cs_wm_common:default_produce_body/2 "
  "riak_cs_s3_response:respond/3 "
  "riak_cs_xml:export_xml/1 "
  "xmerl:export1/3 "   
  "xmerl:export_content/2 "
  "xmerl:export_element/2 "
  "xmerl:export_content/2 "
  "xmerl:tagdef/5 "    
  "xmerl:apply_cb/5 "  
```

**Comment by reiddraper** [comment]
This is bizarre. The only place I see `ensure_loaded` being called is in meck...

**Comment by kellymclaughlin** [comment]
I think I found the culprit. Seems to be the Erlang [error_handler](https://github.com/erlang/otp/blob/maint/lib/kernel/src/error_handler.erl#L39) module. 

Using the following redbug command and running a bucket listing operation with s3cmd:

```
redbug:start("error_handler:undefined_function -> stack", [{msgs,100},{max_msg_size, 9999999999999}]).
```

gives output like this:

```
8:19:36 <{mochiweb_acceptor,init,3}> {error_handler,undefined_function,
                                       [xmerl_xml,'DisplayName',
                                        [["display"],
                                         [],
                                         [{'Owner',6},
                                          {'Contents',8},
                                          {'ListBucketResult',1}],
                                         {xmlElement,'DisplayName',[],[],
                                          {xmlNamespace,[],[]},
                                          [{'Owner',6},
                                           {'Contents',8},
                                           {'ListBucketResult',1}],
                                          2,[],
                                          [{xmlText,
                                            [{'DisplayName',2},
                                             {'Owner',6},
                                             {'Contents',8},
                                             {'ListBucketResult',1}],
                                            1,[],"display",text}],
                                          [],[],undeclared}]]}
  "proc_lib:init_p_do_apply/3 "
  "mochiweb_http:parse_headers/5 "
  "webmachine_mochiweb:loop/2 "
  "webmachine_decision_core:handle_request/2 "
  "webmachine_decision_core:decision/1 "
  "webmachine_decision_core:resource_call/1 "
  "webmachine_resource:do/3 "
  "webmachine_resource:resource_call/3 "
  "riak_cs_wm_common:produce_body/2 "
  "riak_cs_wm_common:default_produce_body/2 "
  "riak_cs_s3_response:respond/3 "
  "riak_cs_xml:export_xml/1 "
  "xmerl:export1/3 "
  "xmerl:export_content/2 "
  "xmerl:export_element/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_element/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:export_element/2 "
  "xmerl:export_content/2 "
  "xmerl:export_content/2 "
  "xmerl:tagdef/5 "
```

Most if not all the output has to do with XML response generation. I am guessing we are horribly misusing xmerl and it is probably my fault. I'll be happy to look into this more tomorrow.

**Comment by shino** [comment]
Call of `xmerl:export_simple` with callback module `xmerl_xml` leads
to the following sequence:
1. for each tag element of the xml doc, `xmerl_xml:'TagName'` is called,
   it is undefined, `error_handler` and `code_server` are called.
2. general callback function `xmerl_xml:'#element#'` is called (as fallback?)
   and iolist of the xml tag element is generated.

I think the current implementation is not misusing, it follows the way
reference document. http://www.erlang.org/doc/man/xmerl.html

A workaround (or proper way??) to avoid undef at the first step
(`xmerl_xml:'TagName'` call) is to implement custom callback module.
A quick patch is: https://gist.github.com/shino/e56b152f285fa89db840

A little measurement on my laptop shows that elaplsed time of
`riak_cs_xml:to_xml` for listing 1000 keys changes from 130msec to 50
msec (custome callback is 80 msec faster than `xmerl_xml`).  But this
approach needs boilerplate callback functions for every XML tag name
(the gist abuses macro :P). So whenever we add XML response with new
tag names, we should add callback functions too.  Any thoughts on the
tradeoff and using the macro?

**Comment by reiddraper** [comment]
This is a big performance win. And I like the macro. Untested, but +1 to this idea from me.

**Comment by kellymclaughlin** [comment]
I opened #674 to address this issue. After reading the `xmerl` source and realizing there was no way to avoid having to implement callbacks for each element name we use, I was frustrated and decided to attempt to extract the `xmerl` source from the rest of OTP and make my own fork. I am happy to hear all opinions on this as a solution, but just wanted to open a PR now to demonstrate the fix and the minimal changes required in the `riak_cs` code.

**Comment by shino** [comment]
I agree that it's better to avoid adding new callback functions for every XML element name.

For another candidate without changing xmerl itself, I wrote #675 , which implements minimal XML exporting functionality needed for riak_cs.

**Comment by shino** [comment]
one more way: modification of xmerl > https://github.com/shino/otp/compare/ss-avoid-undef
- no riak cs modification needed
-  not tested yet :P

**Comment by kellymclaughlin** [comment]
As I have thought about it today, I am in favor of the compromise approach that I mention in [this](https://github.com/basho/riak_cs/pull/674#issuecomment-24609324) comment the PR I made to use my modified `xmerl`. 

I think we should submit a change to xmerl so that all Erlang users can benefit from this. None of the examples that I have found in the `xmerl` documentation use these callbacks and I suspect many people are paying this performance penalty. Assuming the OTP team accepts whatever changes we submit, there is still the problem of the lag time before those changes are reflected in an Erlang release and in one of our releases. So in the near-term I advocate that we use a custom fork to get the performance gains as soon as possible.

I like the look of Shino's `xmerl` change, but I have not tested it yet either. Assuming the performance is about the same with either of our changes, I would be fine going with either one.

**Comment by shino** [comment]
> I think we should submit a change to xmerl so that all Erlang users can benefit from this.

Totally agree.

> I like the look of Shino's xmerl change,

Thank you :) My thought about xmerl is
- interface (API) design to use simple tuples in `xmerl_export_simple` is good
- its implementation which throws a lot of undef errors is bad

So my change respect API and does not change it.
(as a bonus, I can avoid adding test cases :P )

**Comment by kellymclaughlin** [comment]
Let's just make sure the performance using your change is equivalent to using my option. I don't see why it wouldn't be, but we should verify first before submitting a patch. I'm happy to do some testing of it tomorrow.

**Comment by kellymclaughlin** [comment]
> Let's just make sure the performance using your change is equivalent to using my option. I don't see why it wouldn't be, but we should verify first before submitting a patch. I'm happy to do some testing of it tomorrow. 

I spent a few minutes testing this today and both approaches seem to shave 4 or 5 seconds off the time to list a bucket with 30K objects in it. I was using Riak 1.2.1 underneath and with no changes I was seeing around 41 seconds total. With either of the changes that dropped to 36-37 seconds. If you can confirm similar results with your testing then we should submit your patch to OTP and use that approach.

**Comment by shino** [comment]
I also saw improvement for both Kelly's and my changes.
Kelly's one is faster than me :)
- R15B03
- Riak CS development, Riak 1.4.1 (single node on my laptop)
- `fold_object_for_list_keys=true`

Measurement is single threaded, list first 1000 keys, 100 times:

```
$ time zsh -c 'for i in {1..100}; do
      S3CURL=.s3curl.8071.alice s3curl.pl --id cs -- -s
      -x localhost:8071 http://test.s3.amazonaws.com/  > /dev/null;
  done'
```

Results:

```
original
zsh -c   2.38s user 0.70s system 9% cpu 31.208 total

Kelly's
zsh -c   2.37s user 0.77s system 13% cpu 23.353 total

Shino's
zsh -c   2.42s user 0.67s system 12% cpu 24.627 total
```

**Comment by shino** [comment]
PR to Erlang/OTP > https://github.com/erlang/otp/pull/78

**Comment by slfritchie** [comment]
Shino, FYI, the Erlang OTP team doesn't use the GitHub pull request system.  :-(

https://github.com/erlang/otp/wiki/Submitting-patches

**Comment by shino** [comment]
IIRC otp team started using github pull requests. I don't find the information source.

Latest diff of the wiki page shows:
https://github.com/erlang/otp/wiki/Submitting-patches/_compare/96aeaddae63d14ae09a9762e4163a4ea97fa6510%5E...96aeaddae63d14ae09a9762e4163a4ea97fa6510

What I don't understand is whether only PR is sufficient or also email to erlang-patches is needed.
I will ping if there will be no followup.

**Comment by psyeugenic** [comment]
OTP does use GitHubs Pull Request system and is sufficient to send a PR.

**Comment by kuenishi** [comment]
ping; I guess this can be resolved by adopting R16B03 (just ask Jared to use R16B03 when packaging). Any thoughts?

**Comment by shino** [comment]
Yes, no change needed if we can use R16B03.

**Comment by kellymclaughlin** [comment]
Using R16B03 is not sufficient until Riak is also built with the same version because it could cause version issues with the backend module and other code executed as M/R phases. I'd say let's proceed with this for now.

**Comment by shino** [comment]
Good point.
This isuue is addressed by #706 and it is ready for review.

**Comment by shino** [comment]
Addressed by #848
